[test-triage] chip_sw_all_escalation_resets

### Hierarchy of regression failure

Chip Level

### Failure Description

Test chip_sw_all_escalation_resets has 2 failures.
17.chip_sw_all_escalation_resets.1083579376
Log /container/opentitan-public/scratch/os_regression/chip_earlgrey_asic-sim-vcs/17.chip_sw_all_escalation_resets/latest/run.log

      File "/root/.cache/bazel/_bazel_default/96f40d218badc03ed33c48f246ec2aa1/external/rules_rust/rust/private/repository_utils.bzl", line 568, column 20, in load_arbitrary_tool
          ctx.extract(
  Error in extract: java.io.IOException: Error extracting /root/.cache/bazel/_bazel_default/96f40d218badc03ed33c48f246ec2aa1/external/rust_linux_x86_64/rust-1.60.0-x86_64-unknown-linux-gnu.tar.gz to /root/.cache/bazel/_bazel_default/96f40d218badc03ed33c48f246ec2aa1/external/rust_linux_x86_64: Unexpected end of ZLIB input stream
  ERROR: /workspace/mnt/repo_top/sw/host/opentitantool/BUILD:10:12: //sw/host/opentitantool:opentitantool depends on @rust_linux_x86_64//:toolchain_for_x86_64-unknown-linux-gnu_impl in repository @rust_linux_x86_64 which failed to fetch. no such package '@rust_linux_x86_64//': java.io.IOException: Error extracting /root/.cache/bazel/_bazel_default/96f40d218badc03ed33c48f246ec2aa1/external/rust_linux_x86_64/rust-1.60.0-x86_64-unknown-linux-gnu.tar.gz to /root/.cache/bazel/_bazel_default/96f40d218badc03ed33c48f246ec2aa1/external/rust_linux_x86_64: Unexpected end of ZLIB input stream
  ERROR: Analysis of target '//sw/device/tests/sim_dv:all_escalation_resets_test_sim_dv' failed; build aborted:
  INFO: Elapsed time: 704.250s
  INFO: 0 processes.
  FAILED: Build did NOT complete successfully (244 packages loaded, 13962 targets configured)
  FAILED: Build did NOT complete successfully (244 packages loaded, 13962 targets configured)
  make: *** [/workspace/mnt/repo_top/hw/dv/tools/dvsim/sim.mk:73: sw_build] Error 1

### Steps to Reproduce

This test has been failing over the last 5 days but gets to late 90s percentage. 
Sept 12 / Sept 11 / Sept 10 / Sept 9. 0% was previous run on Sept 7 (Missing Sept 8)
GH Commit: https://github.com/lowrisc/opentitan/tree/3c54b1eb225f685fe528a88992a1b8de3e30cd4b
Build seed: 3969941873

Last day with complete failure was https://reports.opentitan.org/hw/top_earlgrey/dv/2022.09.07_23.52.23/report.html

### Tests with similar or related failures

- [ ] chip_sw_clkmgr_escalation_reset
- [ ] chip_sw_pwrmgr_escalation_reset
- [ ] chip_sw_rstmgr_escalation_reset

--- a/sw/device/lib/testing/BUILD
+++ b/sw/device/lib/testing/BUILD
@@ -61,6 +61,7 @@ cc_library(
     deps = [
         "//sw/device/lib/dif:alert_handler",
         "//sw/device/lib/dif:base",
+        "//sw/device/lib/dif:rstmgr",
         "//sw/device/lib/testing/test_framework:check",
     ],
 )

--- a/sw/device/lib/testing/alert_handler_testutils.c
+++ b/sw/device/lib/testing/alert_handler_testutils.c
@@ -2,12 +2,87 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
+#include "sw/device/lib/testing/alert_handler_testutils.h"
+
 #include "sw/device/lib/base/math.h"
 #include "sw/device/lib/dif/dif_alert_handler.h"
 #include "sw/device/lib/dif/dif_base.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
 #include "sw/device/lib/testing/test_framework/check.h"
 
 #include "alert_handler_regs.h"  // Generated
+
+/**
+ * This is used to traverse the dump treating it as an array of bits, and
+ * extract a number of bits placing them in a uint32_t. The word and bit index
+ * are updated by num_bits before returning.
+ */
+static uint32_t get_next_n_bits(
+    int num_bits, const dif_rstmgr_alert_info_dump_segment_t *dump,
+    int *word_index, int *bit_index) {
+  CHECK(num_bits <= 32);
+  CHECK(*bit_index < 32);
+  uint32_t word = dump[*word_index] >> *bit_index;
+  if (*bit_index + num_bits >= 32) {
+    (*word_index) += 1;
+    *bit_index = *bit_index + num_bits - 32;
+  } else {
+    *bit_index += num_bits;
+  }
+  word &= (1 << num_bits) - 1;
+  return word;
+}
+
+alert_info_t alert_info_dump_to_struct(
+    const dif_rstmgr_alert_info_dump_segment_t *dump, int dump_size) {
+  int word_index = 0;
+  int bit_index = 0;
+  alert_info_t info;
+  for (int i = 0; i < ALERT_HANDLER_PARAM_N_CLASSES; ++i) {
+    info.class_esc_state[i] = get_next_n_bits(3, dump, &word_index, &bit_index);
+  }
+  for (int i = 0; i < ALERT_HANDLER_PARAM_N_CLASSES; ++i) {
+    info.class_esc_cnt[i] = get_next_n_bits(32, dump, &word_index, &bit_index);
+  }
+  for (int i = 0; i < ALERT_HANDLER_PARAM_N_CLASSES; ++i) {
+    info.class_accum_cnt[i] =
+        get_next_n_bits(16, dump, &word_index, &bit_index);
+  }
+  info.loc_alert_cause = get_next_n_bits(7, dump, &word_index, &bit_index);
+  CHECK(word_index < dump_size);
+  for (int i = 0; i < ALERT_HANDLER_PARAM_N_ALERTS; ++i) {
+    info.alert_cause[i] = get_next_n_bits(1, dump, &word_index, &bit_index);
+  }
+  CHECK(word_index < dump_size);
+  return info;
+}
+
+void alert_info_to_string(const alert_info_t *info) {
+  LOG_INFO("alert_info:");
+  LOG_INFO("esc_state [0]=%x, [1]=%x, [2]=%x, [3]=%x", info->class_esc_state[0],
+           info->class_esc_state[1], info->class_esc_state[2],
+           info->class_esc_state[3]);
+  LOG_INFO("esc_cnt [0]=0x%x, [1]=0x%x, [2]=0x%x, [3]=0x%x",
+           info->class_esc_cnt[0], info->class_esc_cnt[1],
+           info->class_esc_cnt[2], info->class_esc_cnt[3]);
+  LOG_INFO("accum_cnt [0]=0x%x, [1]=0x%x, [2]=0x%x, [3]=0x%x",
+           info->class_accum_cnt[0], info->class_accum_cnt[1],
+           info->class_accum_cnt[2], info->class_accum_cnt[3]);
+  LOG_INFO("loc_alert_cause=0x%x", info->loc_alert_cause);
+  char cause_bits[81];
+  cause_bits[80] = '\0';
+  int string_index = 0;
+  for (int i = 0; i < ALERT_HANDLER_PARAM_N_ALERTS; ++i) {
+    cause_bits[string_index] = info->alert_cause[i] ? '1' : '0';
+    ++string_index;
+    if (i % 8 == 0) {
+      cause_bits[string_index] = '_';
+      ++string_index;
+    }
+  }
+  cause_bits[string_index] = '\0';
+  LOG_INFO("alert_cause=%s (in increasing order left to right)", cause_bits);
+}
 
 void alert_handler_testutils_configure_all(
     const dif_alert_handler_t *alert_handler, dif_alert_handler_config_t config,
@@ -67,4 +142,8 @@ uint32_t alert_handler_testutils_get_cycles_from_us(uint64_t microseconds) {
         "The value 0x%08x%08x can't fit into the 32 bits timer counter.",
         (cycles >> 32), (uint32_t)cycles);
   return (uint32_t)cycles;
+}
+
+uint32_t alert_handler_testutils_cycle_rescaling_factor() {
+  return kDeviceType == kDeviceSimDV ? 1 : 10;
 }

--- a/sw/device/lib/testing/alert_handler_testutils.h
+++ b/sw/device/lib/testing/alert_handler_testutils.h
@@ -5,8 +5,49 @@
 #ifndef OPENTITAN_SW_DEVICE_LIB_TESTING_ALERT_HANDLER_TESTUTILS_H_
 #define OPENTITAN_SW_DEVICE_LIB_TESTING_ALERT_HANDLER_TESTUTILS_H_
 
+#include <stdbool.h>
+
 #include "sw/device/lib/dif/dif_alert_handler.h"
 #include "sw/device/lib/dif/dif_base.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+
+#include "alert_handler_regs.h"
+
+typedef enum alert_handler_class_state {
+  kCstateIdle = 0,
+  kCstateTimeout = 1,
+  kCstateTerminal = 3,
+  kCstatePhase0 = 4,
+  kCstatePhase1 = 5,
+  kCstatePhase2 = 6,
+  kCstatePhase3 = 7,
+  kCstateFsmError = 2
+} alert_handler_class_state_t;
+
+/**
+ * Represents the hardware alert crash dump in a more software-friendly manner.
+ */
+typedef struct alert_info {
+  bool alert_cause[ALERT_HANDLER_PARAM_N_ALERTS];
+  uint8_t loc_alert_cause;                                  // 7bit
+  uint16_t class_accum_cnt[ALERT_HANDLER_PARAM_N_CLASSES];  // 4x16bit
+  uint32_t class_esc_cnt[ALERT_HANDLER_PARAM_N_CLASSES];    // 4x32bit
+  alert_handler_class_state_t
+      class_esc_state[ALERT_HANDLER_PARAM_N_CLASSES];  // 4x3bit
+} alert_info_t;
+
+/**
+ * Converts the hardware alert crash dump into an alert_info_t.
+ *
+ * This makes it easier to compare and display the different fields.
+ */
+alert_info_t alert_info_dump_to_struct(
+    const dif_rstmgr_alert_info_dump_segment_t *dump, int dump_size);
+
+/**
+ * Displays an alert_info_t as strings.
+ */
+void alert_info_to_string(const alert_info_t *info);
 
 /**
  * Configures alert handler with all required runtime information.
@@ -29,5 +70,23 @@ void alert_handler_testutils_configure_all(
  * Returns the number of cycles corresponding to the given microseconds.
  */
 uint32_t alert_handler_testutils_get_cycles_from_us(uint64_t microseconds);
+
+/**
+ * Returns a scaling factor for conversion of time to cycles.
+ *
+ * Setting escalation cycle counts must be such that the ISRs have time to
+ * complete before the next escalation stage starts. We tend to generate cycle
+ * counts from time durations, and given that the relevant clock frequencies
+ * are much lower in non-sim_dv devices we may end up having to set quite high
+ * time durations to avoid incomplete ISRs for fpga or verilator.
+ *
+ * This function allows the time durations to be set for the sim_dv clock
+ * frequencies, and other platforms get the cycle count rescaled by a factor
+ * of 10. It should be used in the conversion of time duration to cycle counts,
+ * as in
+ *  cycles = udiv64_slow(micros * clockFreqHz, 1000000, NULL) *
+ *           cycle_rescaling_factor();
+ */
+uint32_t alert_handler_testutils_cycle_rescaling_factor();
 
 #endif  // OPENTITAN_SW_DEVICE_LIB_TESTING_ALERT_HANDLER_TESTUTILS_H_

--- a/sw/device/tests/aon_timer_irq_test.c
+++ b/sw/device/tests/aon_timer_irq_test.c
@@ -211,9 +211,9 @@ bool test_main(void) {
       &plic, kPlicTarget, kTopEarlgreyPlicIrqIdAonTimerAonWkupTimerExpired,
       kTopEarlgreyPlicIrqIdAonTimerAonWdogTimerBark);
 
-  // Executing the test using randon time bounds calculated from the clock
+  // Executing the test using random time bounds calculated from the clock
   // frequency to make sure the aon timer is generating the interrupt after the
-  // choosen time and there's no error in the reference time measurement. This
+  // chosen time and there's no error in the reference time measurement. This
   // calculation is required as the various platforms used for testing have
   // differing clocks frequencies. A minimum amount of cycles is required for
   // the interrupt to note the elapsed time so the test fails with an

--- a/sw/device/tests/rstmgr_alert_info_test.c
+++ b/sw/device/tests/rstmgr_alert_info_test.c
@@ -104,17 +104,6 @@ typedef struct node {
   dif_alert_handler_class_t class;
 } node_t;
 
-typedef enum cstate {
-  kCstateIdle = 0,
-  kCstateTimeout = 1,
-  kCstateTerminal = 3,
-  kCstatePhase0 = 4,
-  kCstatePhase1 = 5,
-  kCstatePhase2 = 6,
-  kCstatePhase3 = 7,
-  kCstateFsmError = 2
-} cstate_t;
-
 typedef enum test_round {
   kRound1 = 0,
   kRound2 = 1,
@@ -122,17 +111,6 @@ typedef enum test_round {
   kRound4 = 3,
   kRoundTotal = 4
 } test_round_t;
-
-#define ALERT_CAUSE_CNT ((ALERT_HANDLER_PARAM_N_ALERTS - 1) / 32) + 1
-
-typedef struct alert_info {
-  char *test_name;
-  uint32_t alert_cause[ALERT_CAUSE_CNT];
-  uint8_t loc_alert_cause;                                  // 7bit
-  uint16_t class_accum_cnt[ALERT_HANDLER_PARAM_N_CLASSES];  // 4x16bit
-  uint32_t class_esc_cnt[ALERT_HANDLER_PARAM_N_CLASSES];    // 4x32bit
-  cstate_t class_esc_state[ALERT_HANDLER_PARAM_N_CLASSES];  // 4x3bit
-} alert_info_t;
 
 static volatile test_round_t global_test_round;
 static volatile uint32_t global_alert_called;
@@ -203,106 +181,55 @@ static const dif_alert_handler_class_config_t
             },
 };
 
-static alert_info_t kExpectedInfo[kRoundTotal] = {
+typedef struct test_alert_info {
+  char *test_name;
+  alert_info_t alert_info;
+} test_alert_info_t;
+
+static test_alert_info_t kExpectedInfo[kRoundTotal] = {
     [kRound1] =
         {
             .test_name = "Single class(ClassA)",
-            .class_accum_cnt = {3, 0, 0, 0},
-            .class_esc_state = {kCstatePhase0, kCstateIdle, kCstateIdle,
-                                kCstateIdle},
+            .alert_info =
+                {
+                    .class_accum_cnt = {3, 0, 0, 0},
+                    .class_esc_state = {kCstatePhase0, kCstateIdle, kCstateIdle,
+                                        kCstateIdle},
+                },
         },
     [kRound2] =
         {
             .test_name = "Multi classes(ClassB,C)",
-            .class_accum_cnt = {0, 1, 4, 0},
-            .class_esc_state = {kCstateIdle, kCstatePhase1, kCstatePhase0,
-                                kCstateIdle},
+            .alert_info =
+                {
+                    .class_accum_cnt = {0, 1, 4, 0},
+                    .class_esc_state = {kCstateIdle, kCstatePhase1,
+                                        kCstatePhase0, kCstateIdle},
+                },
         },
     [kRound3] =
         {
             .test_name = "All classes",
-            .class_accum_cnt = {1, 1, 1, 1},
-            .class_esc_state = {kCstatePhase0, kCstatePhase1, kCstatePhase0,
-                                kCstatePhase0},
+            .alert_info =
+                {
+                    .class_accum_cnt = {1, 1, 1, 1},
+                    .class_esc_state = {kCstatePhase0, kCstatePhase1,
+                                        kCstatePhase0, kCstatePhase0},
+                },
         },
     [kRound4] =
         {
             .test_name = "Local alert(ClassB)",
-            .loc_alert_cause = (0x1 << kDifAlertHandlerLocalAlertAlertPingFail),
-            .class_accum_cnt = {0, 1, 0, 0},
-            .class_esc_state = {kCstateIdle, kCstatePhase2, kCstateIdle,
-                                kCstateIdle},
+            .alert_info =
+                {
+                    .loc_alert_cause =
+                        (0x1 << kDifAlertHandlerLocalAlertAlertPingFail),
+                    .class_accum_cnt = {0, 1, 0, 0},
+                    .class_esc_state = {kCstateIdle, kCstatePhase2, kCstateIdle,
+                                        kCstateIdle},
+                },
         },
 };
-
-// TODO: This entire function needs to be made smarter.
-// There is too much hardcoding that changes when we add / subtract
-// alerts.
-// We need to track both a dump_idx and a bit_idx to know where we
-// are and do the conversion in an automatic way.
-static alert_info_t alert_info_dump2struct(
-    const dif_rstmgr_alert_info_dump_segment_t *dump) {
-  // each escalation state is 3 bits, and there are 4 classes total
-  alert_info_t myinfo = {
-      .class_esc_state = {(cstate_t)(dump[0] & 0x7),
-                          (cstate_t)((dump[0] >> 3) & 0x7),
-                          (cstate_t)((dump[0] >> 6) & 0x7),
-                          (cstate_t)((dump[0] >> 9) & 0x7)},
-  };
-
-  // class_esc_cnt
-  uint32_t upper, lower;
-
-  for (int i = 0; i < ALERT_HANDLER_PARAM_N_CLASSES; ++i) {
-    lower = dump[i] >> 12;
-    upper = dump[i + 1] << 20;
-    myinfo.class_esc_cnt[i] = upper | lower;
-  }
-
-  // class_accum_cnt
-  myinfo.class_accum_cnt[0] = (uint16_t)(dump[4] >> 12);
-  lower = dump[4] >> 28;
-  upper = dump[5] << 8;
-  myinfo.class_accum_cnt[1] = (uint16_t)(upper | lower);
-  myinfo.class_accum_cnt[2] = (uint16_t)(dump[5] >> 12);
-  lower = dump[5] >> 28;
-  upper = dump[6] << 8;
-  myinfo.class_accum_cnt[3] = (uint16_t)(upper | lower);
-  // loc_alert_cause
-  myinfo.loc_alert_cause = (dump[6] >> 12) & 0x7f;
-
-  // alert_cause
-  // dump[6] layout:
-  //  31                     19,
-  // |--- alert_cause[12:0]---+
-  // 18                    12,11 0
-  // +--loc_alert_cause[6:0]--+--class_accum_cnt[3][15:4]--|
-  //
-  uint32_t alerts_remaining = kTopEarlgreyAlertIdLast + 1;
-
-  // the lower 32-bit of the alert cause is made up of:
-  // upper 13 bits from dump[6]
-  // lower 19 bits from dump[7]
-  myinfo.alert_cause[0] = dump[6] >> 19;
-  myinfo.alert_cause[0] = myinfo.alert_cause[0] | (dump[7] & ((1 << 20) - 1))
-                                                      << 13;
-  alerts_remaining -= 32;
-
-  // the upper 32-bit of the alert cause is made up of:
-  // upper 13 bits from dump[7]
-  // lower 19 bits from dump[8]
-  myinfo.alert_cause[1] = dump[7] >> 19;
-  myinfo.alert_cause[1] = myinfo.alert_cause[1] | (dump[8] & ((1 << 20) - 1))
-                                                      << 13;
-  alerts_remaining -= 32;
-
-  // shove remaining bits of dump[8] into this one
-  if (alerts_remaining > 0 & alerts_remaining < 32) {
-    myinfo.alert_cause[2] = dump[8] >> 19;
-  }
-
-  return myinfo;
-}
 
 static node_t test_node[kTopEarlgreyAlertPeripheralLast];
 
@@ -383,8 +310,8 @@ void ottf_external_isr(void) {
 }
 
 static void print_alert_cause(alert_info_t info) {
-  for (uint32_t i = 0; i < ALERT_CAUSE_CNT; ++i) {
-    LOG_INFO("alert_cause[%d]: 0x%x", i, (uint32_t)(info.alert_cause[i]));
+  for (uint32_t i = 0; i < ALERT_HANDLER_PARAM_N_ALERTS; ++i) {
+    LOG_INFO("alert_cause[%d]: 0x%x", i, info.alert_cause[i]);
   }
 }
 
@@ -624,42 +551,48 @@ static void collect_alert_dump_and_compare(test_round_t round) {
     LOG_INFO("DUMP:%d: 0x%x", i, dump[i]);
   }
 
-  actual_info = alert_info_dump2struct(dump);
+  actual_info = alert_info_dump_to_struct(dump, seg_size);
 
   if (round == kRound4) {
     // Check local alert only.
     // While testing ping timeout for local alert,
-    // global alert ping timeout can be trigerred
+    // global alert ping timeout can be triggered
     // dut to short timeout value.
     // However, alert source of this ping timeout can be choosen randomly,
-    // as documented in issue #2321, so we conly check local alert cause.
+    // as documented in issue #2321, so we only check local alert cause.
     LOG_INFO("loc_alert_cause: exp: %08x   obs: %08x",
-             kExpectedInfo[round].loc_alert_cause, actual_info.loc_alert_cause);
-    CHECK(kExpectedInfo[round].loc_alert_cause == actual_info.loc_alert_cause);
+             kExpectedInfo[round].alert_info.loc_alert_cause,
+             actual_info.loc_alert_cause);
+    CHECK(kExpectedInfo[round].alert_info.loc_alert_cause ==
+          actual_info.loc_alert_cause);
   } else {
     LOG_INFO("observed alert cause:");
     print_alert_cause(actual_info);
     LOG_INFO("expected alert cause:");
-    print_alert_cause(kExpectedInfo[round]);
-    for (int i = 0; i < ALERT_CAUSE_CNT; ++i) {
-      CHECK(kExpectedInfo[round].alert_cause[i] == actual_info.alert_cause[i]);
+    print_alert_cause(kExpectedInfo[round].alert_info);
+    for (int i = 0; i < ALERT_HANDLER_PARAM_N_ALERTS; ++i) {
+      CHECK(kExpectedInfo[round].alert_info.alert_cause[i] ==
+                actual_info.alert_cause[i],
+            "At alert cause %d Expected %d, got %d", i,
+            kExpectedInfo[round].alert_info.alert_cause[i],
+            actual_info.alert_cause[i]);
     }
   }
 
   for (int i = 0; i < ALERT_HANDLER_PARAM_N_CLASSES; ++i) {
-    CHECK(kExpectedInfo[round].class_accum_cnt[i] ==
+    CHECK(kExpectedInfo[round].alert_info.class_accum_cnt[i] ==
               actual_info.class_accum_cnt[i],
           "alert_info.class_accum_cnt[%d] mismatch exp:0x%x  obs:0x%x", i,
-          kExpectedInfo[round].class_accum_cnt[i],
+          kExpectedInfo[round].alert_info.class_accum_cnt[i],
           actual_info.class_accum_cnt[i]);
   }
   for (int i = 0; i < ALERT_HANDLER_PARAM_N_CLASSES; ++i) {
     // added '<' because expected state can be minimum phase but
     // depends on simulation, sometimes it captures higher phase.
-    CHECK(kExpectedInfo[round].class_esc_state[i] <=
+    CHECK(kExpectedInfo[round].alert_info.class_esc_state[i] <=
               actual_info.class_esc_state[i],
           "alert_info.class_esc_state[%d] mismatch exp:0x%x  obs:0x%x", i,
-          kExpectedInfo[round].class_esc_state[i],
+          kExpectedInfo[round].alert_info.class_esc_state[i],
           actual_info.class_esc_state[i]);
   }
 }
@@ -728,35 +661,32 @@ static node_t test_node[kTopEarlgreyAlertPeripheralLast] = {
 };
 
 static void init_expected_cause() {
-  kExpectedInfo[kRound1].alert_cause[kTopEarlgreyAlertIdI2c0FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdI2c0FatalFault % 32);
-  kExpectedInfo[kRound1].alert_cause[kTopEarlgreyAlertIdI2c1FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdI2c1FatalFault % 32);
-  kExpectedInfo[kRound1].alert_cause[kTopEarlgreyAlertIdI2c2FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdI2c2FatalFault % 32);
+  kExpectedInfo[kRound1]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdI2c0FatalFault] = 1;
+  kExpectedInfo[kRound1]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdI2c1FatalFault] = 1;
+  kExpectedInfo[kRound1]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdI2c2FatalFault] = 1;
 
-  kExpectedInfo[kRound2].alert_cause[kTopEarlgreyAlertIdUart0FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdUart0FatalFault % 32);
-  kExpectedInfo[kRound2].alert_cause[kTopEarlgreyAlertIdUart1FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdUart1FatalFault % 32);
-  kExpectedInfo[kRound2].alert_cause[kTopEarlgreyAlertIdUart2FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdUart2FatalFault % 32);
-  kExpectedInfo[kRound2].alert_cause[kTopEarlgreyAlertIdUart3FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdUart3FatalFault % 32);
   kExpectedInfo[kRound2]
-      .alert_cause[kTopEarlgreyAlertIdOtpCtrlFatalBusIntegError / 32] |=
-      1 << (kTopEarlgreyAlertIdOtpCtrlFatalBusIntegError % 32);
+      .alert_info.alert_cause[kTopEarlgreyAlertIdUart0FatalFault] = 1;
+  kExpectedInfo[kRound2]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdUart1FatalFault] = 1;
+  kExpectedInfo[kRound2]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdUart2FatalFault] = 1;
+  kExpectedInfo[kRound2]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdUart3FatalFault] = 1;
+  kExpectedInfo[kRound2]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdOtpCtrlFatalBusIntegError] = 1;
 
   kExpectedInfo[kRound3]
-      .alert_cause[kTopEarlgreyAlertIdRvCoreIbexRecovSwErr / 32] |=
-      1 << (kTopEarlgreyAlertIdRvCoreIbexRecovSwErr % 32);
-  kExpectedInfo[kRound3].alert_cause[kTopEarlgreyAlertIdUart0FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdUart0FatalFault % 32);
-  kExpectedInfo[kRound3].alert_cause[kTopEarlgreyAlertIdI2c0FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdI2c0FatalFault % 32);
+      .alert_info.alert_cause[kTopEarlgreyAlertIdRvCoreIbexRecovSwErr] = 1;
   kExpectedInfo[kRound3]
-      .alert_cause[kTopEarlgreyAlertIdSpiHost0FatalFault / 32] |=
-      1 << (kTopEarlgreyAlertIdSpiHost0FatalFault % 32);
+      .alert_info.alert_cause[kTopEarlgreyAlertIdUart0FatalFault] = 1;
+  kExpectedInfo[kRound3]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdI2c0FatalFault] = 1;
+  kExpectedInfo[kRound3]
+      .alert_info.alert_cause[kTopEarlgreyAlertIdSpiHost0FatalFault] = 1;
 };
 
 bool test_main(void) {

--- a/sw/device/tests/sim_dv/BUILD
+++ b/sw/device/tests/sim_dv/BUILD
@@ -397,6 +397,7 @@ opentitan_functest(
         "//sw/device/lib/dif:rv_core_ibex",
         "//sw/device/lib/dif:rv_plic",
         "//sw/device/lib/dif:sram_ctrl",
+        "//sw/device/lib/runtime:irq",
         "//sw/device/lib/runtime:log",
         "//sw/device/lib/testing:alert_handler_testutils",
         "//sw/device/lib/testing:aon_timer_testutils",

--- a/sw/device/tests/sim_dv/all_escalation_resets_test.c
+++ b/sw/device/tests/sim_dv/all_escalation_resets_test.c
@@ -49,6 +49,7 @@
 #include "sw/device/lib/testing/test_framework/check.h"
 #include "sw/device/lib/testing/test_framework/ottf_main.h"
 
+#include "alert_handler_regs.h"
 #include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
 
 OTTF_DEFINE_TEST_CONFIG();
@@ -84,19 +85,17 @@ enum {
 };
 
 /**
- * Program the alert handler to escalate on alerts upto phase 2 (i.e. reset).
+ * Program the alert handler to escalate on alerts through NMI and then reset.
  * Also program the aon timer with:
  * - bark after escalation starts, so the interrupt is suppressed by escalation,
  * - bite after escalation reset, so we should not get timer reset.
  */
 enum {
-  kWdogBarkMicros = 2000,          // 2 ms
-  kWdogBiteMicros = 6000,          // 6 ms
-  kEscalationStartMicros = 1000,   // 1 ms
-  kEscalationPhase0Micros = 1000,  // 1 ms
-  kEscalationPhase1Micros = 1000,  // 1 ms
-  kEscalationPhase2Micros = 1000,  // 1 ms
-  kEscalationPhase3Micros = 1000,  // 1 ms
+  kWdogBarkMicros = 200,          // 200 us
+  kWdogBiteMicros = 600,          // 600 us
+  kEscalationStartMicros = 100,   // 100 us
+  kEscalationPhase0Micros = 100,  // 100 us
+  kEscalationPhase1Micros = 100,  // 100 us
   kMaxResets = 2,
   kMaxInterrupts = 30,
 };
@@ -546,23 +545,18 @@ static void alert_handler_config(void) {
   dif_alert_handler_escalation_phase_t esc_phases[] = {
       {.phase = kDifAlertHandlerClassStatePhase0,
        .signal = 0,
-       .duration_cycles = udiv64_slow(
-           kEscalationPhase0Micros * kClockFreqPeripheralHz, 1000000,
-           /*rem_out=*/NULL)},
+       .duration_cycles =
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase0Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()},
       {.phase = kDifAlertHandlerClassStatePhase1,
-       .signal = 1,
-       .duration_cycles = udiv64_slow(
-           kEscalationPhase1Micros * kClockFreqPeripheralHz, 1000000,
-           /*rem_out=*/NULL)},
-      {.phase = kDifAlertHandlerClassStatePhase2,
        .signal = 3,
-       .duration_cycles = udiv64_slow(
-           kEscalationPhase2Micros * kClockFreqPeripheralHz, 1000000,
-           /*rem_out=*/NULL)}};
+       .duration_cycles =
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase1Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()}};
 
   uint32_t deadline_cycles =
-      udiv64_slow(kEscalationStartMicros * kClockFreqPeripheralHz, 1000000,
-                  /*rem_out=*/NULL);
+      alert_handler_testutils_get_cycles_from_us(kEscalationStartMicros) *
+      alert_handler_testutils_cycle_rescaling_factor();
   LOG_INFO("Configuring class A with %d cycles and %d occurrences",
            deadline_cycles, UINT16_MAX);
   dif_alert_handler_class_config_t class_config[] = {{
@@ -856,6 +850,33 @@ static void execute_test(const dif_aon_timer_t *aon_timer) {
   wait_for_interrupt();
 }
 
+void check_alert_dump() {
+  dif_rstmgr_alert_info_dump_segment_t dump[DIF_RSTMGR_ALERT_INFO_MAX_SIZE];
+  size_t seg_size;
+  alert_info_t actual_info;
+
+  CHECK_DIF_OK(dif_rstmgr_alert_info_dump_read(
+      &rstmgr, dump, DIF_RSTMGR_ALERT_INFO_MAX_SIZE, &seg_size));
+
+  LOG_INFO("DUMP SIZE %d", seg_size);
+  for (int i = 0; i < seg_size; i++) {
+    LOG_INFO("DUMP:%d: 0x%x", i, dump[i]);
+  }
+
+  actual_info = alert_info_dump_to_struct(dump, seg_size);
+  LOG_INFO("The alert info crash dump:");
+  alert_info_to_string(&actual_info);
+  // Check alert cause.
+  for (int i = 0; i < ALERT_HANDLER_PARAM_N_ALERTS; ++i) {
+    if (i == kExpectedAlertNumber) {
+      CHECK(actual_info.alert_cause[i], "Expected alert cause %d to be set", i);
+    } else {
+      CHECK(!actual_info.alert_cause[i], "Expected alert cause %d to be clear",
+            i);
+    }
+  }
+}
+
 bool test_main(void) {
   // Enable global and external IRQ at Ibex.
   irq_global_ctrl(true);
@@ -905,6 +926,8 @@ bool test_main(void) {
 
   if (rst_info == kDifRstmgrResetInfoPor) {
     LOG_INFO("Booting for the first time, starting test");
+    // Enable rstmgr alert info capture.
+    CHECK_DIF_OK(dif_rstmgr_alert_info_set_enabled(&rstmgr, kDifToggleEnabled));
     execute_test(&aon_timer);
   } else if (rst_info == kDifRstmgrResetInfoEscalation) {
     restore_fault_checker(&fault_checker);
@@ -912,10 +935,11 @@ bool test_main(void) {
     LOG_INFO("Booting for the second time due to escalation reset");
 
     int interrupt_count = flash_ctrl_testutils_counter_get(kCounterInterrupt);
-    CHECK(interrupt_count > 0, "Expected at least one regular interrupt");
-
     int nmi_interrupt_count = flash_ctrl_testutils_counter_get(kCounterNmi);
-    CHECK(nmi_interrupt_count > 0, "Expected at least one nmi");
+    if (kExpectedAlertNumber != kTopEarlgreyAlertIdFlashCtrlFatalStdErr) {
+      CHECK(interrupt_count > 0, "Expected at least one regular interrupt");
+      CHECK(nmi_interrupt_count > 0, "Expected at least one nmi");
+    }
 
     // Check the alert handler cause is cleared.
     bool is_cause = true;

--- a/sw/device/tests/sim_dv/pwrmgr_deep_sleep_all_reset_reqs_test.c
+++ b/sw/device/tests/sim_dv/pwrmgr_deep_sleep_all_reset_reqs_test.c
@@ -72,10 +72,6 @@ enum {
   kEscalationPhase2Micros = 50,                               // 50 us
 };
 
-uint32_t cycle_rescaling_factor() {
-  return kDeviceType == kDeviceSimDV ? 1 : 10;
-}
-
 static_assert(
     kWdogBarkMicros < kWdogBiteMicros &&
         kWdogBarkMicros > kEscalationPhase0Micros &&
@@ -189,28 +185,24 @@ static void alert_handler_config(void) {
       {.phase = kDifAlertHandlerClassStatePhase0,
        .signal = 0,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase0Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()},
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase0Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()},
       {.phase = kDifAlertHandlerClassStatePhase1,
        .signal = 1,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase1Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()},
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase1Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()},
       {.phase = kDifAlertHandlerClassStatePhase2,
        .signal = 3,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase2Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()}};
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase2Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()}};
 
   dif_alert_handler_class_config_t class_config[] = {{
       .auto_lock_accumulation_counter = kDifToggleDisabled,
       .accumulator_threshold = 0,
-      .irq_deadline_cycles =
-          udiv64_slow(10 * kClockFreqPeripheralHz, 1000000, NULL) *
-          cycle_rescaling_factor(),
+      .irq_deadline_cycles = alert_handler_testutils_get_cycles_from_us(10) *
+                             alert_handler_testutils_cycle_rescaling_factor(),
       .escalation_phases = esc_phases,
       .escalation_phases_len = ARRAYSIZE(esc_phases),
       .crashdump_escalation_phase = kDifAlertHandlerClassStatePhase3,
@@ -241,10 +233,10 @@ static void trigger_escalate(dif_aon_timer_t *aon_timer,
                              const dif_pwrmgr_t *pwrmgr) {
   uint64_t bark_cycles =
       udiv64_slow(kWdogBarkMicros * kClockFreqAonHz, 1000000, NULL) *
-      cycle_rescaling_factor();
+      alert_handler_testutils_cycle_rescaling_factor();
   uint64_t bite_cycles =
       udiv64_slow(kWdogBiteMicros * kClockFreqAonHz, 1000000, NULL) *
-      cycle_rescaling_factor();
+      alert_handler_testutils_cycle_rescaling_factor();
 
   CHECK(bite_cycles < UINT32_MAX,
         "The value %u can't fit into the 32 bits timer counter.", bite_cycles);

--- a/sw/device/tests/sim_dv/pwrmgr_normal_sleep_all_reset_reqs_test.c
+++ b/sw/device/tests/sim_dv/pwrmgr_normal_sleep_all_reset_reqs_test.c
@@ -66,10 +66,6 @@ enum {
   kEscalationPhase2Micros = 50,                               // 50 us
 };
 
-uint32_t cycle_rescaling_factor() {
-  return kDeviceType == kDeviceSimDV ? 1 : 10;
-}
-
 static_assert(
     kWdogBarkMicros < kWdogBiteMicros &&
         kWdogBarkMicros > kEscalationPhase0Micros &&
@@ -183,28 +179,24 @@ static void alert_handler_config(void) {
       {.phase = kDifAlertHandlerClassStatePhase0,
        .signal = 0,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase0Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()},
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase0Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()},
       {.phase = kDifAlertHandlerClassStatePhase1,
        .signal = 1,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase1Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()},
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase1Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()},
       {.phase = kDifAlertHandlerClassStatePhase2,
        .signal = 3,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase2Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()}};
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase2Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()}};
 
   dif_alert_handler_class_config_t class_config[] = {{
       .auto_lock_accumulation_counter = kDifToggleDisabled,
       .accumulator_threshold = 0,
-      .irq_deadline_cycles =
-          udiv64_slow(10 * kClockFreqPeripheralHz, 1000000, NULL) *
-          cycle_rescaling_factor(),
+      .irq_deadline_cycles = alert_handler_testutils_get_cycles_from_us(10) *
+                             alert_handler_testutils_cycle_rescaling_factor(),
       .escalation_phases = esc_phases,
       .escalation_phases_len = ARRAYSIZE(esc_phases),
       .crashdump_escalation_phase = kDifAlertHandlerClassStatePhase3,
@@ -235,10 +227,10 @@ static void config_escalate(dif_aon_timer_t *aon_timer,
                             const dif_pwrmgr_t *pwrmgr) {
   uint64_t bark_cycles =
       udiv64_slow(kWdogBarkMicros * kClockFreqAonHz, 1000000, NULL) *
-      cycle_rescaling_factor();
+      alert_handler_testutils_cycle_rescaling_factor();
   uint64_t bite_cycles =
       udiv64_slow(kWdogBiteMicros * kClockFreqAonHz, 1000000, NULL) *
-      cycle_rescaling_factor();
+      alert_handler_testutils_cycle_rescaling_factor();
 
   CHECK(bite_cycles < UINT32_MAX,
         "The value %u can't fit into the 32 bits timer counter.", bite_cycles);

--- a/sw/device/tests/sim_dv/pwrmgr_random_sleep_all_reset_reqs_test.c
+++ b/sw/device/tests/sim_dv/pwrmgr_random_sleep_all_reset_reqs_test.c
@@ -67,10 +67,6 @@ enum {
   kEscalationPhase2Micros = 50,                               // 50 us
 };
 
-uint32_t cycle_rescaling_factor() {
-  return kDeviceType == kDeviceSimDV ? 1 : 10;
-}
-
 static_assert(
     kWdogBarkMicros < kWdogBiteMicros &&
         kWdogBarkMicros > kEscalationPhase0Micros &&
@@ -184,28 +180,24 @@ static void alert_handler_config(void) {
       {.phase = kDifAlertHandlerClassStatePhase0,
        .signal = 0,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase0Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()},
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase0Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()},
       {.phase = kDifAlertHandlerClassStatePhase1,
        .signal = 1,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase1Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()},
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase1Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()},
       {.phase = kDifAlertHandlerClassStatePhase2,
        .signal = 3,
        .duration_cycles =
-           udiv64_slow(kEscalationPhase2Micros * kClockFreqPeripheralHz,
-                       1000000, NULL) *
-           cycle_rescaling_factor()}};
+           alert_handler_testutils_get_cycles_from_us(kEscalationPhase2Micros) *
+           alert_handler_testutils_cycle_rescaling_factor()}};
 
   dif_alert_handler_class_config_t class_config[] = {{
       .auto_lock_accumulation_counter = kDifToggleDisabled,
       .accumulator_threshold = 0,
-      .irq_deadline_cycles =
-          udiv64_slow(10 * kClockFreqPeripheralHz, 1000000, NULL) *
-          cycle_rescaling_factor(),
+      .irq_deadline_cycles = alert_handler_testutils_get_cycles_from_us(10) *
+                             alert_handler_testutils_cycle_rescaling_factor(),
       .escalation_phases = esc_phases,
       .escalation_phases_len = ARRAYSIZE(esc_phases),
       .crashdump_escalation_phase = kDifAlertHandlerClassStatePhase3,
@@ -236,10 +228,10 @@ static void config_escalate(dif_aon_timer_t *aon_timer,
                             const dif_pwrmgr_t *pwrmgr) {
   uint64_t bark_cycles =
       udiv64_slow(kWdogBarkMicros * kClockFreqAonHz, 1000000, NULL) *
-      cycle_rescaling_factor();
+      alert_handler_testutils_cycle_rescaling_factor();
   uint64_t bite_cycles =
       udiv64_slow(kWdogBiteMicros * kClockFreqAonHz, 1000000, NULL) *
-      cycle_rescaling_factor();
+      alert_handler_testutils_cycle_rescaling_factor();
 
   CHECK(bite_cycles < UINT32_MAX,
         "The value %u can't fit into the 32 bits timer counter.", bite_cycles);

